src/main/java/io/usethesource/vallang/IValue.java

```diff
@@ -28,7 +28,35 @@ public interface IValue {
 	 * @return the {@link Type} of a value
 	 */
     public Type getType();
+
+    /**
+     * This method is used exclusively by code generated by the Rascal compiler,
+     * or by the Rascal interpreter. The returned integer codes are opaque, although stable.
+     * If you need to know what kind of value you have, use the IValueVisitor or the ITypeVisitor
+     * interfaces and the `accept` methods on IValue and Type.
+     * 
+     * @return an integer code that:
+     *    * accurate reflects the identity of the top-level structure of this value
+     *    * such that if pattern.match(this) ===> pattern.getPatternMatchFingerprint() == this.getPatternMatchFingerprint()
+     *    * distinguishes maximally between different kinds of values
+     *    * never makes the same or similar value have a different fingerprint
+     */
+    default int getMatchFingerprint() {
```

Comment:
Can you add a remark on the stability of these ints across different JVM instances/invocations? Or between different jvm versions?

Reply:
See the tests. They document the stability in several ways. Changes are not allowed at all. If different VM 's lead to changed codes, I'm trying to let the tests fail.

Reply:
This is a huge part of the binary backwards compatibility of generated Rascal java code. If the ints change here, all the previously compiled rascal code breaks against the new fingerprints. It would do that in ridiculous ways by calling the wrong overloads, visiting the wrong case, or failing switch and dispatches altogether without warning.

Reply:
The issue of different hash functions in different JVMs is indeed troublesome. Why don't we remove hash codes completely from this story and opt for wisely chosen constants.

Reply:
the hash codes are stable acros implementations of JVM's, as we implement them mostly ourselves. We have copies and extensions of the String.hashCode code all over the places. Also, the JVM switches itself on those codes for strings, so they can not change them any more than we can.

Still I made the tests so that we will see any change immediately.

* I have made all implementation that return constants, to simply return that constant.  So for lists you get a number that is currently equal to "list".hashCode(); etc.
* for literal values we use hashCodes; since it is quite hard to come up with a better spread than what hashCode already is doing. We can either completely copy that code instead of calling hashCode for reuse, or simply call hashCode until it breaks and we have to copy the old implementation in getMatchFingerprint.

Reply:
for `String.hashCode` it's indeed documented now, so that likely won't change (and like you said, it's used for the switch in java).

As long as we don't use the hashCode of other base objects, we should be fine.